File: min_swaps.py
testes = [
    [0, 0, 0, 0, 0, 0],
    [1, 1, 1, 0, 0, 0],
    [0, 0, 0, 1, 1, 1],
    [1, 0, 1, 0, 1, 0],
    [0, 1, 1, 0, 1, 0],
    [0, 0, 1, 0, 0, 0],
    [0, 0, 0, 1, 0, 0],
    [0, 0, 1, 1, 1, 0],
]


def swap(lst, i, j):
    print(f'({i}, {j}): {lst} -> ', end='')
    lst[i], lst[j] = lst[j], lst[i]
    print(f'{lst}')


def sort(lst, comparison):
    moves = 0
    for i in range(len(lst)-1):
        if comparison(lst[i], lst[i+1]):
            for j in range(i, -1, -1):
                if lst[j+1] == lst[j]:
                    break
                moves += 1
                swap(lst, j, j+1)
    print('Finish:', lst)
    print('Moves:', moves, '\n')
    return moves


if __name__ == '__main__':

    for t in testes:
        print('Start:', t, '\n')
        minimum = min(sort(t.copy(), lambda a, b: a > b),
                      sort(t.copy(), lambda a, b: a < b))
        print(f'Minimum moves: {minimum}')
        print(30*'-', '\n', sep='')

File: min_swaps.hs
{- What is the minimum amount of swaps between neighbors
    to group all equal elements on a list with 2 groups ({0,1})? -}

-- 0,0,0,3,4,2,2,3
testes = [[0, 0, 0, 0, 0, 0],
          [1, 1, 1, 0, 0, 0],
          [0, 0, 0, 1, 1, 1],
          [1, 0, 1, 0, 1, 0],
          [0, 1, 1, 0, 1, 0],
          [0, 0, 1, 0, 0, 0],
          [0, 0, 0, 1, 0, 0],
          [0, 0, 1, 1, 1, 0]]


push :: Ord a => [a] -> [a] -> (a -> a -> Bool) -> Int -> (Int, [a])
push ref []  _ count = (count, reverse ref)
push ref [a] _ count = (count, reverse (a:ref))
push ref (x:y:xs) comp count
    | comp x y = push ref (y:x:xs) comp (count+1)
    | otherwise = push (x:ref) (y:xs) comp count


-- count and past_count MUST start with different values
sprintInternal :: Ord a => (a -> a -> Bool) -> [a] -> Int -> Int -> (Int, [a])
sprintInternal comp lst count past_count
    | count == past_count = (count, nlist)
    | count == 0 = sprintInternal comp nlist ncount count
    | otherwise = sprintInternal comp nlist (ncount+count) count
    where (ncount, nlist) = push [] lst comp 0


sprint :: Ord a => (a -> a -> Bool) -> [a] -> (Int, [a])
sprint comp lst = sprintInternal comp lst 0 (-1)


main = do
    let menor = map (sprint (<)) testes
    let maior = map (sprint (>)) testes
    print $ zipWith min menor maior
